File: Heap/Q3. Construct Target Array With Multiple Sums/Q3_ConstructTargetArrayWithMultipleSums.py
class Solution(object):
    def isPossible(self, target):
        """
        :type target: List[int]
        :rtype: bool
        """
        heap = [-num for num in target]        # Create max heap using negative values
        total = sum(target)                    # Store sum of all elements
        heapify(heap)                          # Convert list into a heap

        while heap[0] != -1:                   # Continue until all values become 1
            num = -heappop(heap)               # Get the largest value
            total -= num                       # Remove it from total sum

            if num <= total or total < 1:      # Check if operation is invalid
                return False                   # Not possible to reach target

            num %= total                       # Reduce the largest value
            total += num                       # Add updated value to total

            heappush(heap, -num or -total)     # Push updated value back into heap

        return True                            # Target array is possible

File: Heap/Q3. Construct Target Array With Multiple Sums/Q3_ConstructTargetArrayWithMultipleSums.java
import java.util.Collections;

public boolean isPossible(int[] target) {          // Check if target array can be built
    if (target.length == 1)                        // If only one element
        return target[0] == 1;                     // It must be 1

    PriorityQueue<Integer> pq =
        new PriorityQueue<>(Collections.reverseOrder()); // Max heap to get largest value
    int sum = 0;                                   // Store sum of all elements

    for (int t : target) {                         // Loop through target array
        pq.add(t);                                 // Add element to heap
        sum += t;                                  // Add element to total sum
    }

    while (pq.peek() != 1) {                       // Continue until largest is 1
        int curr = pq.poll();                      // Remove largest element

        if (sum - curr == 1)                       // If remaining sum is 1
            return true;                           // Target is possible

        int x = curr % (sum - curr);               // Reduce current value
        sum = sum - curr + x;                      // Update total sum

        if (x == 0 || x == curr)                   // If no valid change
            return false;                          // Not possible
        else
            pq.add(x);                             // Add updated value back to heap
    }

    return true;                                   // All elements reduced to 1
}
